cmd/aws_oidc.go

```
@@ -1,14 +1,53 @@
 package cmd
 
 import (
+	"encoding/json"
 	"github.com/aws/aws-sdk-go/aws"
 	"github.com/aws/aws-sdk-go/aws/session"
 	"github.com/aws/aws-sdk-go/service/sts"
 	"github.com/pkg/errors"
+	"io/ioutil"
+	"os"
+	"time"
 )
 
+const awsCredentialsCache = "aws_token.json"
+const expiryDelta = 10 * time.Second
+
+type AWSCredentials struct {
+	AWSAccessKey     string
+	AWSSecretKey     string
+	AWSSessionToken  string
+	AWSSecurityToken string
+	PrincipalARN     string
+	Expires          time.Time
+}
+
+func (cred AWSCredentials) Valid() bool {
+	if cred.Expires.IsZero() {
+		return false
+	}
+	return !cred.Expires.Add(-expiryDelta).Before(time.Now())
+}
+
 func GetCredentialsWithOIDC(client *OIDCClient, idToken string, durationSeconds int64) (*AWSCredentials, error) {
-	return loginToStsUsingIDToken(client, idToken, durationSeconds)
+	jsonBytes, err := ioutil.ReadFile(awsCredentialsCache)
+	var awsCreds *AWSCredentials = nil
+	if !os.IsNotExist(err) {
```

Please let the program crash on unhandled errors